src/supervisor/message_handlers.py

```diff
@@ -163,6 +171,239 @@ def _build_gzip_attachment(self, input_path):  #pylint: disable=no-self-use
             return attachment
 
 
+class SendGridHandler(MessageHandler):  # pylint: disable=too-few-public-methods
+    """Send emails via the SendGrid service.
+
+    Attributes
+    ----------
+    sendgrid_settings : dict
+        'from_address' (str)
+        'to_addresses' (str or List): single string or list of strings
+        'api_key' (str): SendGrid api key
+    project_name : str
+        pip-install name of the client project for client version number reporting
+
+    Methods
+    -------
+    send_message(message_details)
+        Build a message and send using the SendGrid API's helper classes
+    """
+
+    def __init__(self, sendgrid_settings, project_name=''):
+        self.sendgrid_settings = sendgrid_settings
+        self.sendgrid_client = sendgrid.SendGridAPIClient(api_key=self.sendgrid_settings['api_key'])
+        self.project_name = project_name
+
+    def send_message(self, message_details):
+        """Construct and send an email message with the SendGrid API
+
+        Args:
+            message_details : MessageDetails
+                Must have .message, .subject; may have .attachments
+        """
+
+        from_address, to_addresses = self._verify_addresses()
+        #: Bail out instead of raising error instead of forcing client to deal with it
+        #: Warnings should have been raised from _verify_addresses() and Nones returned.
+        if not from_address or not to_addresses:
+            return
+
+        sender_address = Email(from_address)
+        recipient_addresses = self._build_recipient_addresses(to_addresses)
+
+        subject = self._build_subject(message_details)
+        attachment_warning, verified_attachments = self._verify_attachments(message_details.attachments)
+        new_message = attachment_warning + message_details.message
+        content = self._build_content(new_message, self.project_name)
+        attachments = self._process_attachments(verified_attachments)
+
+        #: Build message object and send it
+        mail = Mail(sender_address, recipient_addresses, subject, content)
+        mail.attachment = attachments
+        response = self.sendgrid_client.client.mail.send.post(request_body=mail.get())  # pylint: disable=unused-variable
+
+        #: Maybe test the response via response.status_code?
```

Comment:
**chore**: I definitely think that you should verify that the email was sent successfully before you return out of this method.

Reply:
Have either of you had any success interrogating these responses? I just keep getting `202` from `response.status_code`, even if I put in a non-existent recipient address. Oddly, this isn't even listed in their [documented status codes](https://docs.sendgrid.com/api-reference/how-to-use-the-sendgrid-v3-api/responses). If I try to send a mal-formatted `Mail` object it throws an error with a 404 code in the message, so a try/except may make sense here. However, I've not yet found any way to get a "we tried sending it but it didn't get delivered" status.

Reply:
I don't have any experience with it. Sounds like it may be fine as is if there isn't a way to fail sending a message without the code raising an exception.